File: src/domain/entities/issue.ts
import { IssueId } from '../value-objects/issue-id.js';
import { IssueStatus } from '../value-objects/issue-status.js';
import { IssueType } from '../value-objects/issue-type.js';
import { ProjectId } from '../value-objects/project-id.js';

import type { TimeProvider } from '../interfaces/time-provider.js';

export interface IssueSnapshot {
  id: string;
  title: string;
  description: string;
  type: string;
  status: string;
  projectId?: string;
  parentId?: string;
  childIds: string[];
  dependencies: string[];
  estimate?: number;
  createdAt: Date;
  updatedAt: Date;
}

export class Issue {
  private _id: IssueId;
  private _title: string;
  private _description: string;
  private _type: IssueType;
  private _status: string;
  private _projectId?: ProjectId;
  private _parentId?: IssueId;
  private _childIds: IssueId[];
  private _dependencies: IssueId[];
  private _estimate?: number;
  private _createdAt: Date;
  private _updatedAt: Date;

  private static readonly MAX_TITLE_LENGTH = 255;
  private static readonly FIBONACCI_SEQUENCE = [1, 2, 3, 5, 8, 13, 21, 34, 55, 89];

  constructor(
    id: IssueId,
    title: string,
    description: string,
    type: string,
    status: string,
    childIds: IssueId[],
    dependencies: IssueId[],
    parentId: IssueId | undefined,
    projectId: ProjectId | undefined,
    estimate: number | undefined,
    createdAt: Date,
    updatedAt: Date,
    private readonly timeProvider?: TimeProvider
  ) {
    this._id = id;
    this._title = this.validateTitle(title);
    this._description = description;
    this._type = type;
    this._status = status;
    if (projectId !== undefined) {
      this._projectId = projectId;
    }
    if (parentId !== undefined) {
      this._parentId = parentId;
    }
    this._childIds = [...childIds];
    this._dependencies = [...dependencies];
    if (estimate !== undefined) {
      this._estimate = estimate;
    }
    this._createdAt = new Date(createdAt);
    this._updatedAt = new Date(updatedAt);
  }

  get id(): IssueId {
    return this._id;
  }

  get title(): string {
    return this._title;
  }

  get description(): string {
    return this._description;
  }

  get type(): IssueType {
    return this._type;
  }

  get status(): string {
    return this._status;
  }

  get projectId(): ProjectId | undefined {
    return this._projectId;
  }

  get parentId(): IssueId | undefined {
    return this._parentId;
  }

  get childIds(): IssueId[] {
    return [...this._childIds];
  }

  get dependencies(): IssueId[] {
    return [...this._dependencies];
  }

  get estimate(): number | undefined {
    return this._estimate;
  }

  get createdAt(): Date {
    return new Date(this._createdAt);
  }

  get updatedAt(): Date {
    return new Date(this._updatedAt);
  }

  static create(
    title: string,
    description: string,
    type: string,
    timeProvider?: TimeProvider,
    projectId?: ProjectId
  ): Issue {
    const id = IssueId.generate();
    const now = timeProvider?.now() ?? new Date();
    
    return new Issue(
      id,
      title,
      description,
      type,
      IssueStatus.Backlog,
      [],
      [],
      undefined,
      projectId,
      undefined,
      now,
      now,
      timeProvider
    );
  }

  static fromSnapshot(
    snapshot: IssueSnapshot,
    timeProvider?: TimeProvider
  ): Issue {
    const id = IssueId.from(snapshot.id);
    const projectId = snapshot.projectId ? ProjectId.from(snapshot.projectId) : undefined;
    const parentId = snapshot.parentId ? IssueId.from(snapshot.parentId) : undefined;
    const childIds = snapshot.childIds.map(childId => IssueId.from(childId));
    const dependencies = snapshot.dependencies.map(depId => IssueId.from(depId));
    
    return new Issue(
      id,
      snapshot.title,
      snapshot.description,
      snapshot.type,
      snapshot.status,
      childIds,
      dependencies,
      parentId,
      projectId,
      snapshot.estimate,
      snapshot.createdAt,
      snapshot.updatedAt,
      timeProvider
    );
  }

  toSnapshot(): IssueSnapshot {
    const snapshot: IssueSnapshot = {
      id: this._id.toString(),
      title: this._title,
      description: this._description,
      type: this._type as string,
      status: this._status,
      childIds: this._childIds.map(child => child.toString()),
      dependencies: this._dependencies.map(dep => dep.toString()),
      createdAt: this._createdAt,
      updatedAt: this._updatedAt
    };
    
    if (this._projectId !== undefined) {
      snapshot.projectId = this._projectId.toString();
    }
    
    if (this._parentId !== undefined) {
      snapshot.parentId = this._parentId.toString();
    }
    
    if (this._estimate !== undefined) {
      snapshot.estimate = this._estimate;
    }
    
    return snapshot;
  }

  setParent(parentId: IssueId): void {
    if (this._type === IssueType.Epic) {
      throw new Error('Epic cannot have a parent');
    }
    
    this._parentId = parentId;
    this.touch();
  }

  addChild(childId: IssueId): void {
    if (this._type === IssueType.Subtask) {
      throw new Error('Subtask cannot have children');
    }

    if (!this._childIds.some(id => id.equals(childId))) {
      this._childIds.push(childId);
      this.touch();
    }
  }

  removeChild(childId: IssueId): void {
    const index = this._childIds.findIndex(id => id.equals(childId));

    if (index !== -1) {
      this._childIds.splice(index, 1);
      this.touch();
    }
  }

  hasChild(childId: IssueId): boolean {
    return this._childIds.some(id => id.equals(childId));
  }

  childCount(): number {
    return this._childIds.length;
  }

  addDependency(dependencyId: IssueId): void {
    if (dependencyId.equals(this._id)) {
      throw new Error('Issue cannot depend on itself');
    }

    if (!this._dependencies.some(id => id.equals(dependencyId))) {
      this._dependencies.push(dependencyId);
      this.touch();
    }
  }

  removeDependency(dependencyId: IssueId): void {
    const index = this._dependencies.findIndex(id => id.equals(dependencyId));

    if (index !== -1) {
      this._dependencies.splice(index, 1);
      this.touch();
    }
  }

  hasDependency(dependencyId: IssueId): boolean {
    return this._dependencies.some(id => id.equals(dependencyId));
  }

  dependencyCount(): number {
    return this._dependencies.length;
  }

  updateStatus(newStatus: string): void {
    if (this._status === newStatus) {
      return;
    }

    if (!IssueStatus.canTransition(this._status, newStatus)) {
      throw new Error(`Invalid status transition from ${this._status} to ${newStatus}`);
    }

    this._status = newStatus;
    this.touch();
  }

  setEstimate(estimate: number): void {
    if (this._type === IssueType.Epic) {
      throw new Error('Cannot set estimate on Epic');
    }

    if (this._type === IssueType.Story && this._childIds.length > 0) {
      throw new Error('Cannot set estimate on Story with children');
    }

    if (estimate <= 0) {
      throw new Error('Estimate must be positive');
    }

    if (!Issue.FIBONACCI_SEQUENCE.includes(estimate)) {
      throw new Error('Estimate must follow Fibonacci sequence (1, 2, 3, 5, 8, 13, etc.)');
    }

    this._estimate = estimate;
    this.touch();
  }

  clearEstimate(): void {
    delete this._estimate;
    this.touch();
  }

  updateTitle(title: string): void {
    this._title = this.validateTitle(title);
    this.touch();
  }

  updateDescription(description: string): void {
    this._description = description;
    this.touch();
  }

  updateType(type: IssueType): void {
    if (this._type === type) {
      return;
    }
    
    // Validate type change rules
    if (this._type === IssueType.Epic && this._parentId) {
      throw new Error('Cannot change Epic with parent to another type');
    }
    
    if (type === IssueType.Epic && this._parentId) {
      throw new Error('Cannot change to Epic when issue has a parent');
    }
    
    if (type === IssueType.Subtask && !this._parentId) {
      throw new Error('Cannot change to Subtask without a parent');
    }
    
    this._type = type;
    this.touch();
  }

  isValidHierarchy(): boolean {
    if (this._type === IssueType.Epic) {
      return this._parentId === undefined;
    }
    
    if (this._type === IssueType.Subtask) {
      return this._parentId !== undefined;
    }
    
    return true;
  }

  isBlocked(): boolean {
    return this._dependencies.length > 0;
  }

  isActive(): boolean {
    return IssueStatus.isActive(this._status);
  }

  isCompleted(): boolean {
    return IssueStatus.isCompleted(this._status);
  }

  private validateTitle(title: string): string {
    const trimmedTitle = title.trim();
    
    if (!trimmedTitle) {
      throw new Error('Issue title cannot be empty');
    }

    if (trimmedTitle.length >= Issue.MAX_TITLE_LENGTH) {
      throw new Error('Issue title must be less than 255 characters');
    }

    return trimmedTitle;
  }

  private touch(): void {
    this._updatedAt = this.timeProvider?.now() ?? new Date();
  }
}

File: src/main/kotlin/io/spiralhouse/cycletime/domain/entities/Issue.kt
package io.spiralhouse.cycletime.domain.entities

import io.spiralhouse.cycletime.domain.exceptions.DomainException
import io.spiralhouse.cycletime.domain.services.TimeProvider
import io.spiralhouse.cycletime.domain.valueobjects.*
import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable

/**
 * Immutable snapshot of an Issue entity's state.
 * Used for persistence and serialization without exposing internal mutability.
 */
@Serializable
data class IssueSnapshot(
    val id: IssueId,
    val title: String,
    val description: String?,
    val type: IssueType,
    val status: IssueStatus,
    val parentId: IssueId?,
    val projectId: ProjectId?,
    val estimate: Estimate,
    val assigneeId: String?,
    val dependencies: List<IssueId>,
    val blockedBy: List<IssueId>,
    val createdAt: Instant,
    val updatedAt: Instant,
    val deletedAt: Instant? = null
)

/**
 * Domain entity representing an Issue with rich business logic.
 *
 * An Issue represents work items in the system with support for hierarchy,
 * status transitions, estimation, and dependency management.
 *
 * This entity follows Domain-Driven Design principles with:
 * - Factory methods for controlled instantiation
 * - Encapsulated state with business rule enforcement
 * - Snapshot pattern for persistence
 */
class Issue private constructor(
    val id: IssueId,
    private var _title: String,
    private var _description: String?,
    val type: IssueType,
    private var _status: IssueStatus,
    val parentId: IssueId?,
    val projectId: ProjectId?,
    private var _estimate: Estimate,
    private var _assigneeId: String?,
    private val _dependencies: MutableSet<IssueId>,
    private val _blockedBy: MutableSet<IssueId>,
    val createdAt: Instant,
    private var _updatedAt: Instant,
    private val _deletedAt: Instant?,
    private val timeProvider: TimeProvider
) {
    val title: String get() = _title
    val description: String? get() = _description
    val status: IssueStatus get() = _status
    val estimate: Estimate get() = _estimate
    val assigneeId: String? get() = _assigneeId
    val dependencies: List<IssueId> get() = _dependencies.toList()
    val blockedBy: List<IssueId> get() = _blockedBy.toList()
    val updatedAt: Instant get() = _updatedAt
    val deletedAt: Instant? get() = _deletedAt

    /**
     * Updates the issue status following valid transition rules.
     */
    fun updateStatus(newStatus: IssueStatus) {
        if (!_status.canTransitionTo(newStatus)) {
            throw DomainException("Cannot transition from ${_status.name} to ${newStatus.name}")
        }
        _status = newStatus
        updateTimestamp()
    }

    /**
     * Updates the issue title.
     */
    fun updateTitle(newTitle: String) {
        validateTitle(newTitle)
        _title = newTitle
        updateTimestamp()
    }

    /**
     * Updates the issue description.
     */
    fun updateDescription(newDescription: String?) {
        _description = newDescription
        updateTimestamp()
    }

    /**
     * Updates the issue assignee.
     */
    fun updateAssignee(newAssigneeId: String?) {
        _assigneeId = newAssigneeId
        updateTimestamp()
    }

    /**
     * Sets the estimate for this issue.
     */
    fun setEstimate(newEstimate: Estimate) {
        // Validate estimate is appropriate for issue type
        when (type) {
            IssueType.EPIC -> {
                if (newEstimate.hasValue()) {
                    throw DomainException("Epics cannot have estimates")
                }
            }
            IssueType.STORY -> {
                // Stories can have estimates when they have no subtasks
                // This would require checking for subtasks in a full implementation
            }
            IssueType.SUBTASK -> {
                if (!newEstimate.hasValue()) {
                    throw DomainException("Subtasks must have estimates")
                }
            }
        }

        _estimate = newEstimate
        updateTimestamp()
    }

    /**
     * Adds a dependency (this issue depends on the specified issue).
     */
    fun addDependency(dependencyId: IssueId) {
        if (dependencyId == id) {
            throw DomainException("Issue cannot depend on itself")
        }
        if (_dependencies.contains(dependencyId)) {
            throw DomainException("Dependency already exists")
        }
        _dependencies.add(dependencyId)
        updateTimestamp()
    }

    /**
     * Removes a dependency.
     */
    fun removeDependency(dependencyId: IssueId) {
        _dependencies.remove(dependencyId)
        updateTimestamp()
    }

    /**
     * Adds an issue that blocks this one.
     */
    fun addBlockedBy(blockerId: IssueId) {
        if (blockerId == id) {
            throw DomainException("Issue cannot be blocked by itself")
        }
        _blockedBy.add(blockerId)
        updateTimestamp()
    }

    /**
     * Removes a blocking issue.
     */
    fun removeBlockedBy(blockerId: IssueId) {
        _blockedBy.remove(blockerId)
        updateTimestamp()
    }

    /**
     * Checks if this issue has any dependencies.
     */
    fun hasDependencies(): Boolean = _dependencies.isNotEmpty()

    /**
     * Checks if this issue is blocked by other issues.
     */
    fun isBlocked(): Boolean = _blockedBy.isNotEmpty()

    /**
     * Checks if this issue can be estimated based on its type.
     */
    fun canBeEstimated(): Boolean = when (type) {
        IssueType.EPIC -> false
        IssueType.STORY -> true // In real implementation, would check for subtasks
        IssueType.SUBTASK -> true
    }

    /**
     * Validates hierarchy rules for this issue.
     */
    fun validateHierarchy(parentType: IssueType?) {
        if (!type.canHaveParent(parentType)) {
            val parentName = parentType?.name ?: "no parent"
            throw DomainException("${type.name} cannot have parent of type $parentName")
        }
    }

    /**
     * Creates an immutable snapshot of the current issue state.
     */
    fun toSnapshot(): IssueSnapshot {
        return IssueSnapshot(
            id = id,
            title = _title,
            description = _description,
            type = type,
            status = _status,
            parentId = parentId,
            projectId = projectId,
            estimate = _estimate,
            assigneeId = _assigneeId,
            dependencies = _dependencies.toList(),
            blockedBy = _blockedBy.toList(),
            createdAt = createdAt,
            updatedAt = _updatedAt,
            deletedAt = _deletedAt
        )
    }

    private fun updateTimestamp() {
        _updatedAt = timeProvider.now()
    }

    companion object {
        private const val MAX_TITLE_LENGTH = 255

        /**
         * Factory method to create a new Issue.
         */
        fun create(
            title: String,
            description: String?,
            type: IssueType,
            parentId: IssueId? = null,
            projectId: ProjectId? = null,
            timeProvider: TimeProvider?
        ): Issue {
            requireNotNull(timeProvider) { "TimeProvider cannot be null" }
            validateTitle(title)

            // Validate hierarchy - deferred full validation to application layer
            // The domain entity performs basic validation only. Full hierarchy validation
            // (e.g., verifying parent types) requires repository access and should be
            // implemented in the application service layer where cross-aggregate
            // validation can be performed.
            //
            // Current validation:
            // - Epics cannot have parents (enforced here)
            // - Subtasks should have parents (recommendation, not enforced)
            // - Stories can optionally have Epic parents (not enforced here)
            if (parentId != null && type == IssueType.EPIC) {
                throw DomainException("Epics cannot have parents")
            }

            val now = timeProvider.now()
            return Issue(
                id = IssueId.generate(),
                _title = title,
                _description = description,
                type = type,
                _status = IssueStatus.BACKLOG,
                parentId = parentId,
                projectId = projectId,
                _estimate = Estimate.none(),
                _assigneeId = null,
                _dependencies = mutableSetOf(),
                _blockedBy = mutableSetOf(),
                createdAt = now,
                _updatedAt = now,
                _deletedAt = null,
                timeProvider = timeProvider
            )
        }

        /**
         * Reconstitutes an Issue from a snapshot.
         */
        fun fromSnapshot(
            snapshot: IssueSnapshot,
            timeProvider: TimeProvider?
        ): Issue {
            requireNotNull(timeProvider) { "TimeProvider cannot be null" }

            return Issue(
                id = snapshot.id,
                _title = snapshot.title,
                _description = snapshot.description,
                type = snapshot.type,
                _status = snapshot.status,
                parentId = snapshot.parentId,
                projectId = snapshot.projectId,
                _estimate = snapshot.estimate,
                _assigneeId = snapshot.assigneeId,
                _dependencies = snapshot.dependencies.toMutableSet(),
                _blockedBy = snapshot.blockedBy.toMutableSet(),
                createdAt = snapshot.createdAt,
                _updatedAt = snapshot.updatedAt,
                _deletedAt = snapshot.deletedAt,
                timeProvider = timeProvider
            )
        }

        private fun validateTitle(title: String) {
            when {
                title.isBlank() -> {
                    throw DomainException("Issue title cannot be empty")
                }
                title.length > MAX_TITLE_LENGTH -> {
                    throw DomainException("Issue title cannot exceed $MAX_TITLE_LENGTH characters")
                }
            }
        }
    }
}
